keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1107 @@
+reconciliation of related leader leases).
+
+Coordinated Election Controller reconciliation loop:
+
+- If no leader lease exists for a components:
+  - Elect leader from candidates by preparing a freshly renewed lease with:
+    - `spec.holderIdentity` set to the identity of the elected leader
+    - `coordination.k8s.io/elected-by: leader-election-controller` (to make
+      lease types easy to disambiguate)
+- If there is a better candidate than current leader:
+  - Sets `coordination.k8s.io/end-of-term: true` on the leader lease, signaling
+    that the leader should stop renewing the lease and yield leadership
+
+```mermaid
+flowchart TD
+   A[Reconcile] --> |Process Leader Lease| B
+   B{Lease Status?} --> |Better Leader Exists| D
+   B --> |Expired/Missing| E
+   D[End Lease Term]
+   E[Elect Leader]
+```
+
+Example of a lease created by Coordinated Election Controller:
+
+```yaml
+apiVersion: coordination.k8s.io/v1
+kind: Lease
+metadata:
+  annotations:
+    coordination.k8s.io/elected-by: coordinated-election-controller
+  name: sample-controller
+  namespace: kube-system
+spec:
+  holderIdentity: controller-a
+  leaseDurationSeconds: 10
+  leaseTransitions: 0
+  renewTime: "2023-12-05T18:58:31.295467Z"
+```
+
+The Coordinated Election Controller will run in the kube-apiserver.
+
+The apiserver runs very few controllers, and they are not elected, but instead
+all run concurrently in HA configurations.  Because of this, controller in the
+apiserver must make careful use of concurrency control primitives to ensure
+multiple instances collaborate, not fight. This is discussed in depth in the
+below design details.
+
+### Coordinated Lease Lock
+
+A new `resourceLock` type of `coordinatedleases`, and `CoordinatedLeaseLock`
+implementation of `resourcelock.Interface` will be added to client-go that:
+
+- Creates Identity Lease when ready to be Leader
+	- Renews identity lease periodically
+- Watches Leader Lease, waiting to be elected leader by the Coordinated Election Controller
+- When it becomes leader:
+  - Perform role of active component instance
+  - Renew leader lease periodically
+  - Stop renewing if lease is marked `coordination.k8s.io/end-of-term: true`
+- If leader lease expires:
+  - Shutdown (yielding leadership) and restart as a candidate component instance
+
+```mermaid
+flowchart TD
+    A[Started] -->|Create Identity Lease| B
+    B[Candidate] --> |Elected| C[Leader]
+    C --> |Renew Leader Lease| C
+    C -->|End of Term / Leader Lease Expired| D[Shutdown]
+    D[Shutdown] -.-> |Restart| A
+```
+
+### Enabling on a component
+
+Components with a `--leader-elect-resource-lock` flag (kube-controller-manager,
+ kube-scheduler) will accept `coordinatedleases` as a resource lock type.
+
+### Migrations
+
+So long as the API server is running a coordinated election controller, it is
+safe to directly migrate a component from Lease Based Leader Election to
+Coordinated Leader Election (or vis-versa).
```

Comment:
just to double check,

kubeadm today does not follow the upgrade skew policy correctly.
it has a bit of a legacy mechanism.

what happens during upgrade on a control plane node today is that all control plane components receive new configuration and are restarted by the kubelet (as satatic pods), not in a sequence but randomly, determining by the pod sanbox in the kubelet.

in such a scenario, on the first CP node, a KCM and/or a kube-scheduler could be with the `--leader-elect-resource-lock=coordinatedleases` configuration opted-in already started, yet not a single kube-apiserver instance will support the coordinated LE yet in this HA setup, because e.g. the new kube-apiserver on that same node starts slower.

i assume that would not be a problem?

Reply:
I believe with our current approach KCM/kube-scheduler will not be able to be leader until the new kube-apiserver instance is ready. KCM will be unable to write its identity lease and will not be eligible for leader. This is similar to how it today though right? If kube-apiserver is starting and KCM cannot read the Lease object, they will not be able to make a decision on being the leader.

In an HA environment, this is unaffected since during an upgrade of one control plane instance, another healthy control plane KCM/kube-scheduler will take over.